File: pkg/embedded/identity.go
// Package embedded implements the v1 embedded identity token contract.
// Identity is not a resource grant. Callers must check current authorization.
package embedded

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"strings"
	"time"
)

const TTL = time.Minute
const TokenType = "emisell-embedded-id+jwt"

var ErrInvalid = errors.New("invalid embedded identity")

type Identity struct {
	MerchantID     string `json:"merchantId"`
	ActorID        string `json:"sub"`
	AppID          string `json:"appId"`
	InstallationID string `json:"installationId"`
}

type Claims struct {
	Identity
	Issuer    string `json:"iss"`
	Audience  string `json:"aud"`
	IssuedAt  int64  `json:"iat"`
	ExpiresAt int64  `json:"exp"`
	ID        string `json:"jti"`
}

type header struct {
	Algorithm string `json:"alg"`
	Type      string `json:"typ"`
	KeyID     string `json:"kid"`
}

func valid(v string) bool {
	if len(v) == 0 || len(v) > 256 {
		return false
	}
	for _, r := range v {
		if r < 33 || r > 126 {
			return false
		}
	}
	return true
}

func (i Identity) Valid() bool {
	return valid(i.MerchantID) && valid(i.ActorID) && valid(i.AppID) && valid(i.InstallationID)
}

// Issue must only be called after current staff membership, installation,
// approved launch binding and grant have been authorized server-side.
func Issue(key ed25519.PrivateKey, keyID, issuer, audience string, id Identity, now time.Time) (string, error) {
	if len(key) != ed25519.PrivateKeySize || !valid(keyID) || !valid(issuer) || !valid(audience) || !id.Valid() || now.Unix() <= 0 {
		return "", ErrInvalid
	}
	nonce := make([]byte, 24)
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}
	h, _ := json.Marshal(header{"EdDSA", TokenType, keyID})
	c, _ := json.Marshal(Claims{id, issuer, audience, now.Unix(), now.Add(TTL).Unix(), base64.RawURLEncoding.EncodeToString(nonce)})
	input := base64.RawURLEncoding.EncodeToString(h) + "." + base64.RawURLEncoding.EncodeToString(c)
	return input + "." + base64.RawURLEncoding.EncodeToString(ed25519.Sign(key, []byte(input))), nil
}

func decode(s string, dst any) error {
	b, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil || base64.RawURLEncoding.EncodeToString(b) != s {
		return ErrInvalid
	}
	d := json.NewDecoder(bytes.NewReader(b))
	d.DisallowUnknownFields()
	if d.Decode(dst) != nil {
		return ErrInvalid
	}
	var extra any
	if d.Decode(&extra) != io.EOF {
		return ErrInvalid
	}
	return nil
}

// Verify pins issuer, app-client audience, identity and locally trusted keys.
// It never follows token-supplied URLs. Signature validity does not establish
// current membership or grant: use the application's current-access check too.
func Verify(token string, keys map[string]ed25519.PublicKey, issuer, audience string, expected Identity, now time.Time) (Claims, error) {
	var c Claims
	if len(token) > 4096 || !expected.Valid() || !valid(issuer) || !valid(audience) {
		return c, ErrInvalid
	}
	p := strings.Split(token, ".")
	if len(p) != 3 {
		return c, ErrInvalid
	}
	var h header
	if decode(p[0], &h) != nil || h.Algorithm != "EdDSA" || h.Type != TokenType || !valid(h.KeyID) {
		return c, ErrInvalid
	}
	k := keys[h.KeyID]
	sig, err := base64.RawURLEncoding.DecodeString(p[2])
	if err != nil || len(k) != ed25519.PublicKeySize || !ed25519.Verify(k, []byte(p[0]+"."+p[1]), sig) {
		return c, ErrInvalid
	}
	if decode(p[1], &c) != nil || c.Identity != expected || c.Issuer != issuer || c.Audience != audience || c.IssuedAt <= 0 || c.IssuedAt > now.Unix() || c.ExpiresAt <= now.Unix() || c.ExpiresAt-c.IssuedAt != int64(TTL/time.Second) || len(c.ID) != 32 {
		return Claims{}, ErrInvalid
	}
	return c, nil
}

File: packages/developer-cli/templates/embedded/server/identity.mjs
import { createPublicKey, verify } from 'node:crypto';

export class IdentityError extends Error {
  constructor(code) { super(code); this.code = code; }
}
const invalid = () => { throw new IdentityError('invalid_identity'); };
const valid = value => typeof value === 'string' && /^[\x21-\x7e]{1,256}$/.test(value);
const identityKeys = ['merchantId', 'sub', 'appId', 'installationId'];
function exactKeys(value, keys) {
  return value && typeof value === 'object' && !Array.isArray(value) &&
    Object.keys(value).length === keys.length && keys.every(k => Object.hasOwn(value, k));
}
function decode(value) {
  if (!/^[A-Za-z0-9_-]+$/.test(value)) invalid();
  const bytes = Buffer.from(value, 'base64url');
  if (bytes.toString('base64url') !== value) invalid();
  return bytes;
}
function json(value) {
  try { return JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(decode(value))); }
  catch { invalid(); }
}

// All configuration and callbacks are trusted backend inputs, never browser input.
// resolveExpected must independently resolve installation/staff context from your
// trusted integration. Claims are only lookup hints, never authorization evidence.
// authorizeCurrent must check current installation, staff, client, release and grant.
export function createIdentityVerifier({ keys, issuer, audience, resolveExpected, authorizeCurrent, now = () => Date.now() } = {}) {
  const trusted = new Map();
  if (keys && typeof keys === 'object') for (const [kid, pem] of Object.entries(keys)) {
    if (!valid(kid)) throw new IdentityError('verifier_not_configured');
    const key = createPublicKey(pem);
    if (key.asymmetricKeyType !== 'ed25519') throw new IdentityError('verifier_not_configured');
    trusted.set(kid, key);
  }
  return async function verifySession(token, { signal } = {}) {
    if (!trusted.size || !valid(issuer) || !valid(audience) || typeof resolveExpected !== 'function' || typeof authorizeCurrent !== 'function') throw new IdentityError('verifier_not_configured');
    if (typeof token !== 'string' || token.length > 4096) invalid();
    const parts = token.split('.');
    if (parts.length !== 3) invalid();
    const header = json(parts[0]);
    if (!exactKeys(header, ['alg', 'typ', 'kid']) || header.alg !== 'EdDSA' || header.typ !== 'emisell-embedded-id+jwt' || !valid(header.kid)) invalid();
    const key = trusted.get(header.kid), signature = decode(parts[2]);
    if (!key || signature.length !== 64 || !verify(null, Buffer.from(parts[0] + '.' + parts[1]), key, signature)) invalid();
    const claims = json(parts[1]);
    const timeValid = () => {
      const clock = Math.floor(now() / 1000);
      return Number.isSafeInteger(clock) && Number.isSafeInteger(claims.iat) && Number.isSafeInteger(claims.exp) && claims.iat > 0 && claims.iat <= clock && claims.exp > clock && claims.exp - claims.iat === 60;
    };
    if (!exactKeys(claims, [...identityKeys, 'iss', 'aud', 'iat', 'exp', 'jti']) ||
        !identityKeys.every(k => valid(claims[k])) || claims.iss !== issuer || claims.aud !== audience ||
        typeof claims.jti !== 'string' || !/^[A-Za-z0-9_-]{32}$/.test(claims.jti) || !timeValid()) invalid();
    signal?.throwIfAborted();
    const hints = Object.freeze(Object.fromEntries(identityKeys.map(k => [k, claims[k]])));
    const expected = await resolveExpected(hints, { audience, signal });
    if (!expected || !identityKeys.every(k => valid(expected[k]) && expected[k] === claims[k])) throw new IdentityError('access_denied');
    // Require an explicit boolean; an empty/no-op async callback must not grant access.
    if (await authorizeCurrent(hints, { audience, signal }) !== true) throw new IdentityError('access_denied');
    signal?.throwIfAborted();
    if (!timeValid()) invalid();
    return { merchantId: claims.merchantId, actorId: claims.sub, appId: claims.appId, installationId: claims.installationId, expiresAt: claims.exp };
  };
}
